File: MTGLibs/src/main/java/com/common/control/base/BaseAdapter.java
package com.common.control.base;

import android.content.Context;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

abstract public class BaseAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    protected OnActionCallback callback;
    protected List<T> mList;
    protected Context context;

    public BaseAdapter(List<T> mList, Context context) {
        this.mList = mList;
        this.context = context;
    }

    public void setCallback(OnActionCallback callback) {
        this.callback = callback;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return viewHolder(parent, viewType);
    }

    protected abstract RecyclerView.ViewHolder viewHolder(ViewGroup parent, int viewType);


    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        onBindView(viewHolder, position);
    }

    protected abstract void onBindView(RecyclerView.ViewHolder viewHolder, int position);

    @Override
    public int getItemCount() {
        if (mList != null) {
            return mList.size();
        } else {
            return 0;
        }
    }


}

File: app/src/main/java/com/alx/findphone/claptofind/flashalert/base/BaseAdapter.kt
package com.alx.findphone.claptofind.flashalert.base

import android.content.Context
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.common.control.base.OnActionCallback

abstract class BaseAdapter<T> : RecyclerView.Adapter<RecyclerView.ViewHolder> {
    var mCallback: OnActionCallback? = null
    var mList: List<T> = ArrayList()
    var context: Context? = null

    constructor()
    constructor(mList: List<T>) {
        this.mList = mList
    }

    constructor(mList: List<T>, context: Context?) {
        this.mList = mList
        this.context = context
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        context = parent.context
        return viewHolder(parent, viewType)
    }

    protected abstract fun viewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder
    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        onBindView(viewHolder, position)
    }

    protected abstract fun onBindView(viewHolder: RecyclerView.ViewHolder?, position: Int)
    override fun getItemCount(): Int {
        return mList.size
    }

    fun submitList(mList: List<T>) {
        this.mList = mList
        notifyDataSetChanged()
    }
}